app/src/db/bulk_ops.py

```diff
@@ -0,0 +1,150 @@
+"""Bulk database operations for performance.
+
+Provides a bulk_upsert function
+"""
+from typing import Any, Optional, Sequence
+
+import psycopg
+from psycopg import rows, sql
+
+Connection = psycopg.Connection
+Cursor = psycopg.Cursor
+kwargs_row = rows.kwargs_row
+
+
+def bulk_upsert(
+    cur: psycopg.Cursor,
+    table: str,
+    attributes: Sequence[str],
+    objects: Sequence[Any],
+    constraint: str,
+    update_condition: Optional[sql.SQL] = None,
```

Comment:
The newer form for optional/nullable fields saves an import.
```suggestion
    update_condition: sql.SQL | None = None,
```

Reply:
TIL, ty!